homeassistant/components/tplink/switch.py

```diff
@@ -32,131 +34,50 @@ async def async_setup_entry(
-        """Turn the LED switch off."""
-        await self._led_module.set_led(False)
-
-    @callback
-    def _async_update_attrs(self) -> None:
-        """Update the entity's attributes."""
-        self._attr_is_on = self._led_module.led
-
-    @callback
-    def _handle_coordinator_update(self) -> None:
-        """Handle updated data from the coordinator."""
-        self._async_update_attrs()
-        super()._handle_coordinator_update()
-
-
-class SmartPlugSwitch(CoordinatedTPLinkEntity, SwitchEntity):
-    """Representation of a TPLink Smart Plug switch."""
-
-    _attr_name: str | None = None
+class Switch(CoordinatedTPLinkEntity, SwitchEntity):
+    """Representation of a feature-based TPLink sensor."""
 
     def __init__(
         self,
         device: Device,
         coordinator: TPLinkDataUpdateCoordinator,
+        feature: Feature,
+        parent: Device | None = None,
     ) -> None:
         """Initialize the switch."""
-        super().__init__(device, coordinator)
-        # For backwards compat with pyHS100
-        self._attr_unique_id = legacy_device_id(device)
-        self._async_update_attrs()
-
-    @async_refresh_after
-    async def async_turn_on(self, **kwargs: Any) -> None:
-        """Turn the switch on."""
-        await self.device.turn_on()
-
-    @async_refresh_after
-    async def async_turn_off(self, **kwargs: Any) -> None:
-        """Turn the switch off."""
-        await self.device.turn_off()
-
-    @callback
-    def _async_update_attrs(self) -> None:
-        """Update the entity's attributes."""
-        self._attr_is_on = self.device.is_on
+        super().__init__(device, coordinator, feature=feature, parent=parent)
+        self._feature: Feature
 
-    @callback
-    def _handle_coordinator_update(self) -> None:
-        """Handle updated data from the coordinator."""
-        self._async_update_attrs()
-        super()._handle_coordinator_update()
-
-
-class SmartPlugSwitchChild(SmartPlugSwitch):
-    """Representation of an individual plug of a TPLink Smart Plug strip."""
+        # Use the device name for the primary switch control
+        if feature.category is Feature.Category.Primary:
+            self._attr_name = None
 
-    def __init__(
-        self,
-        device: Device,
-        coordinator: TPLinkDataUpdateCoordinator,
-        plug: Device,
-    ) -> None:
-        """Initialize the child switch."""
-        self._plug = plug
-        super().__init__(device, coordinator)
-        self._attr_unique_id = legacy_device_id(plug)
-        self._attr_name = plug.alias
+        self.entity_description = _description_for_feature(
+            SwitchEntityDescription, feature
+        )
 
     @async_refresh_after
     async def async_turn_on(self, **kwargs: Any) -> None:
-        """Turn the child switch on."""
-        await self._plug.turn_on()
+        """Turn the switch on."""
+        await self._feature.set_value(True)  # type: ignore[no-untyped-call]
```

Comment:
We need to fix this upstream, but it's fine in this development branch to unblock other platforms that depend on the `entity.py` updates done in this PR. Related upstream PR: https://github.com/python-kasa/python-kasa/pull/926

Reply:
It would be better to fix it before merging so we don't have to analyze this code again when we merge back to core

Reply:
I think this might not get fixed for this release as the solution for typing the feature values in the library is still up in the air and we might not be able to do it for all use cases. Can we go forwards with this as library limitation that may or may not get resolved?

Reply:
Alternatively, we could just type with with a union of all values we currently have? `int | float | str | Enum | None` should probably cover it.

Reply:
Seems fine to do an overly broad type for now and than narrow it down later when it could be improved or using a typevar